Add trending statuses

Gargron opened the pull request:
Follow-up to #16917, related issue #4870

A status will appear trending based on the amount of reblogs it got (decaying from the time the status was published), if the status is public, the author has opted-in to be discoverable, the status is not sensitive or hidden behind a content warning, and a moderator has previously reviewed the account (or the individual status) and allowed it to trend. Conversely, even if an account has been allowed to trend, this can be overridden for individual statuses.

![grafik](https://user-images.githubusercontent.com/184731/152662812-ddfffc55-0a9e-45b3-9b8a-d303cc31a828.png)

Furthermore, trending links and trending statuses are now classified by locale, and logged-in users get results in their selected locale. And instead of sending a separate notification e-mail for each type of trend, a single e-mail with all items that need reviewing is sent at most once every 2 hours.

### REST API changes

- `GET /api/v1/trends/statuses` added

### Other

- Remove unused column `trust_level` on accounts table
- Move trend-related filters and batch forms under the `Trends` namespace
- Change unique accounts requirement for trending links from 15 back to 5
- Change review threshold rank from 10 back to 3 for trending tags and trending links

Gargron committed:
Add trending statuses

Gargron committed:
Merge branch 'main' into feature-trending-statuses

ClearlyClaire reviewed (changes requested):
I really don't think we should do this.

## My opinion on the feature itself

More seriously, before I start with the technical review itself, let me say that I am uncomfortable with this feature. This continues what I think constitutes a *trend* to value engagement and the number of daily active users with choices that may be detrimental to the quality of the interactions or the actual experience of the end-user.

Indeed, while user retention can be a useful metric to figure out if the UX is too confusing or tedious, for instance, I think it is fine for people to try out Mastodon and figure out that this kind of microblogging platform is, in fact, not what they were looking for.

Similarly, I understand that trending hashtags and posts (sorry, I still don't understand trending articles) can be useful to discover content and users who may interest you, but… are you doing that to help people find what they were initially searching for, or to constantly give them something new to engage with?

Trending posts in particular likely worsen the “popularity contest” side of things (together with an additional incentive for bots boosting and faving content to be promoted) and require proactive moderation.

On that note, *a priori* moderation likely goes a long way towards preventing the kind of hellish trends of some other networks, but it is likely to cause a sizeable increase in moderation workload, and the implementation still is partial *a priori* moderation, as you can approve accounts rather than individual posts.

## A point on the PR process

This PR is yet another gigantic commit that not only adds trending posts but also:
- performs an undocumented change to the linter configuration
- cleans up unrelated stuff in the `Account` model
- changes the namespace of pre-existing trend stuff
- refactors some of the trend querying code
- changes trend notifications to generate one mail per batch, instead of one mail per batch per trending content type
- adds admin API for pre-existing trend stuff
- changes thresholds for existing trend stuff
- changes trending links to be classified by locale

Some of that is listed in the PR description as “Other”, but honestly that should have been other PRs.

## The technical review itself

I have found few technical issues with the PR, but one thing that stands out is the `Trend::Query` refactor. It could have been good, but I think the mutable nature of the `Query` object (especially given its functional-like interface, and how it's used throughout) is extremely confusing and error-prone.

On a design perspective:
- as for enabling the feature and notifications, I still don't think tying all the trend types together makes a lot of sense (see #17040)
- I find questionable to reduce trending toots to their first line and links to the attachments on the very page were they are supposed to be reviewed, also the layout will probably not look great with more than one attachment there

Something somewhat unrelated I just noticed when reviewing the view templates: I think we're using “•” and “·” in an inconsistent way across the admin and settings view as well as the Web UI.

Gargron committed:
Fix dangling items with stale scores in localized sets

Gargron commented on .rubocop.yml:
```diff
@@ -32,10 +32,11 @@ Layout/EmptyLineAfterGuardClause:
+Layout/FirstHashElementIndentation:
+  EnforcedStyle: consistent
+
 Layout/HashAlignment:
   Enabled: false
-  # EnforcedHashRocketStyle: table
-  # EnforcedColonStyle: table
```
This makes it so you can write

```ruby
foo = {
  bar: 123,
}
```

Without having to align the hash to the `foo =`

Gargron commented on app/models/status.rb:
```diff
@@ -256,6 +257,18 @@ def decrement_count!(key)
+    end
+  end
+
+  def requires_review_notification?
+    attributes['trendable'].nil? && account.requires_review_notification?
+  end
```
Yes, I think that's the correct balance of moderation effort. The second line of defence (as before) is that anything rule-breaking would get reported.

Gargron committed:
Various fixes and improvements

- Change approve_all/reject_all to approve_accounts/reject_accounts
- Change Trends::Query methods to not mutate the original query
- Change Trends::Query#skip to offset
- Change follow recommendations to be refreshed in a transaction

Gargron commented:
Addressed points about trends query mutability. Discovered and fixed bug that left dangling items with stale scores in localized subsets.

Gargron committed:
Add tests for trending statuses filtering behaviour

Gargron committed:
Fix not applying filtering scope in controller

Gargron committed:
Merge remote-tracking branch 'origin/main' into feature-trending-statuses

ClearlyClaire commented on app/models/trends/statuses.rb:
```diff
@@ -0,0 +1,142 @@
+
+    original_status.public_visibility? &&
+      original_status.account.discoverable? && !original_status.account.silenced? &&
+      original_status.spoiler_text.blank? && !original_status.sensitive? && !original_status.reply?
```
unlike trending links, where the CW is lost and thus leading to lost context, having CWed toots trending is probably fine; in fact, filtering out CWed toots for trendings my be an intensive to *not* use the feature

Gargron replied:
We can't easily differentiate between CWs that are there because it's food and the author is conscious of people who don't want to see food, and CWs that are there because it's adult content. We err on the safe side here. Besides, browsing things you can't see and need to tap on is not the most satisfying experience.

trwnh replied:
> We can't easily differentiate between CWs that are there because it's food and the author is conscious of people who don't want to see food, and CWs that are there because it's adult content. 

This is a bit tangential, but would you be open to a feature request to split the "CW" feature from the "mark as sensitive" feature? I know they're at least two separate fields in ActivityPub and also in the API, but currently setting a CW forces the status to be marked sensitive.

ClearlyClaire replied:
> We can't easily differentiate between CWs that are there because it's food and the author is conscious of people who don't want to see food, and CWs that are there because it's adult content.

CWs are not just about “food” or “adult content”, but can cover anxiety-inducing topics or give a very concise summary of the topic of a long toot. Many legitimate and important toots have Content Warnings. Also, some instances may want “adult content” to trend.

> We err on the safe side here. Besides, browsing things you can't see and need to tap on is not the most satisfying experience.

Being exposed to anxiety-inducing or triggering content without a proper warning is not the most satisfying experience either, and I'd argue that not covering those topics at all could be considered as unsatisfying as well, but I guess that depends on your definition.

progval replied:
Two other things to consider:

1. if toots with a CW cannot trend, then it will encourage some/most people not to use CWs because they want their toots to be popular
2. new users who browse trending toots will get used to not seeing CWs, so they won't use CWs themselves when tooting; which will lead overall to people not using CWs.

ClearlyClaire reviewed (commented):
I'm still unconvinced the feature as a whole is a good idea, I still think it should be possible to enable individual trending features independently, or at least notifications for them, and I still find it dubious to reduce would-be  trending toots to their first line and links to the attachments on the very page were they are supposed to be reviewed. Furthermore, I think excluding toots with content warnings from the trending toots might not be a good idea.

Otherwise, the code seems ok.

Gargron commented:
I appreciate your review. I expect that we will adjust things over time.
